### ISSUE TITLE
[QUESTION]: Learning Rates/lr scheduler

### ISSUE THREAD

**Issue by smoes-prime** [chemprop/chemprop#259]
I am using the MPNN in chemprop to generate features and then make predictions. I am trying to optimize the the performance of the MPNN by changing various flag when I run the training, and I noticed there are three different flags for learning rates:
**final_lr
init_lr
max_lr**
and then there is:
**num_lrs**: int
    The number of learning rates to use (currently hard-coded to 1).

so I have interpreted this to mean that init_lr is then the only learning rate used for training the model, and that a learning rate scheduler might be coming in a future update perhaps?

My question is whether or not init_lr is the actual learning rate or is one of the other two used in some way.

Thanks in advance for any help!

**Comment by davidegraff** [comment]
per the paper ["Attention is all you need"](https://arxiv.org/pdf/1706.03762.pdf), the Noam learning rate scheduler:
1. starts the learning rate, `lr` at `init_lr`
2. linearly increases `lr` to `max_lr` over `warmup_epochs` epochs
3. exponentially decays `lr` to `final_lr` as `epoch` tends to `max_epochs`

It does this for each individual `param_group` inside the `Optimizer` passed into the initializer, meaning that at any point in time there could be `num_lrs` independent `lr`'s that the optimizer is keeping track of. An example of this would be an `Adam` optimizer that's optimizing two different parts of your model differently, e.g., the `Encoder` and the `FFN` of the `chemprop` model

**Comment by cjmcgill** [comment]
@davidegraff the names of the arguments generally seem okay. A short description in the readme could help perhaps. 

Though `num_lr` as an internal variable could use a new name to make it clear that it’s about subdivisions of parameters (unused) rather than the number of steps it uses.

**Comment by davidegraff** [comment]
Yeah I have no issue with the variable naming- they're all clear except for `num_lrs`. Why not just call it `num_param_groups` since that's what it *physically* represents? This is also the most minor of nitpicks because it's all internal as well. In fact, I'm not even sure how clients can be confused by this variable, seeing as they don't actually see it... I certainly don't care to spend the effort opening up a PR for that...

**Comment by cjmcgill** [comment]
I think it’s a good change but a low priority. The next time we touch learning rate (maybe for expanded hyperopt options), we can include an update.

**Comment by cjmcgill** [comment]
@smoes-prime , does the explanation of the parameters answer your question?

**Comment by smoes-prime** [comment]
Yes, thanks!
